client/extensions/hello-dolly/state/reducer.js

```diff
@@ -0,0 +1,30 @@
+/**
+ * Internal dependencies
+ */
+import lyrics from './lyrics';
+import {
+	HELLO_DOLLY_NEXT_LYRIC,
+} from './action-types';
+
+const initialState = {
+	lyricIndex: 0
+};
+
+export default function( state = initialState, action ) {
+	switch ( action.type ) {
+		case HELLO_DOLLY_NEXT_LYRIC:
+			return advanceToNextLyric( state );
+		default:
+			return state;
+	}
+}
```

Comment:
what's the purpose of storing an object here as state when it looks like all we're actually storing is a number? objects as state obscure the structure of the tree and make for complicated reducers. what if we just made initial state a primitive data type?

```js
export const lyricIndex = ( state = 0, { type } ) =>
	HELLO_DOLLY_NEXT_LYRIC === type
		? ( lyricIndex + 1 ) % lyrics.length
		: state;
```

I think it's easy to forget how simple and straightforward reducers can be. Redux works extremely well via composition (in this case with `combineReducers`) such that we can extend our state by joining together independent parts instead of adding new things into old things (storing objects as state and adding new properties).

For example, in `advanceToNextLyric()` it looks like we're leaving the door open to new future properties of this state object, but I would ask why we need or want that. If we have a new and different piece of data, it should get its own reducer entirely.

> Redux pushes us to think "what changes this data?" instead of "what data does this change?"